include/nuttx/mm/vm_map.h

```diff
@@ -0,0 +1,176 @@
+/****************************************************************************
+ * include/nuttx/mm/vm_map.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __MM_VM_MAP_VM_MAP_H
+#define __MM_VM_MAP_VM_MAP_H
+
+#ifdef CONFIG_MM_VM_MAP
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/fs/fs.h>
+#include <queue.h>
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* Type of the mapping */
+
+enum vm_map_type
+{
+  VM_MAP_ANONYMOUS,
+  VM_MAP_NAMED,
+  VM_MAP_FILE,
+  VM_MAP_SHM,
+  VM_MAP_RAMMAP,
```

Comment:
Since rammap and shm is implemented now, I would like this patchset can harmonize shm/rammap and file, other could be added later.

Reply:
I can't test the rammap and shm atm, so I wouldn't really like to tinker with those. 

Initially it looks like g_shminfo handling could be replaced with this same vm lists, as well as g_rammaps. But those are both really global lists (not task group / address environment specific) and I fail to understand how those interfaces even work in the first place. I think it would need a bit more study and testing on how to do that.

IMHO, it is not a good thing to put too much into a same PR, and I don't see a problem of having those existing ones as parallel for time being

Reply:
> I can't test the rammap and shm atm, so I wouldn't really like to tinker with those.
> 
> Initially it looks like g_shminfo handling could be replaced with this same vm lists, as well as g_rammaps. But those are both really global lists (not task group / address environment specific)

g_shminfo/g_rammaps need be global, because the same memory block may be shared by multiple process.

> and I fail to understand how those interfaces even work in the first place. I think it would need a bit more study and testing on how to do that.

The equivalent of vm_map_s(tg_vm_map) is group_shm_s(task_group_s::tg_shm).

> 
> IMHO, it is not a good thing to put too much into a same PR, and I don't see a problem of having those existing ones as parallel for time being

Reply:
Please be more specific, what does it mean to "unify the file and ram backend" ?

If suggestion is to implement a generic file mmap, that I won't do:
1) It doesn't make any sense in flat or protected build, since the whole file contents would need to be allocated into ram at mmap
2) For kernel build this would require at least allocating new pages in page fault, and automatically dropping pages when needed -> in practice implementing full virtual memory, which again doesn't make sense in real-time os context
3) I don't need such, so I am not implementing it

I don't see any difference in operation of TMPFS/ROMFS, and I didn't touch those. What kind of problem do you see caused by this?

Reply:
> Please be more specific, what does it mean to "unify the file and ram backend" ?
> 
> If suggestion is to implement a generic file mmap, that I won't do:
> 
> 1. It doesn't make any sense in flat or protected build, since the whole file contents would need to be allocated into ram at mmap

It isn't always true. For example, you can directly map virtual address to file content without allocation in case of ROMFS+XIP Flash. This can even work with more complex file system, if the filesystem allocate the block in the same page size. FIOC_MMAP is designed for this special case.

> 2. For kernel build this would require at least allocating new pages in page fault, and automatically dropping pages when needed -> in practice implementing full virtual memory, which again doesn't make sense in real-time os context

Yes, what I mean the unification is the management data(e.g. per-task/global data list, api entry point), the difference is in the implementation policy which could be guarded by Kconfig option.

> 3. I don't need such, so I am not implementing it
> 

I know your intent is to support the memory mapping with the ram backend(like the current shm?). But, the same mechanism should work with TMPFS and ROMFS. So, it's good to ensure your change could:

1. Unify the ram backend implementation(either base on your new shm driver or shm_info_s g_shminfo, and remove another).
2. The kernel mode addition could work with ROMFS/TMPFS too.

> I don't see any difference in operation of TMPFS/ROMFS, and I didn't touch those. What kind of problem do you see caused by this?

So your virtual mapping work with all(SHM driver, TMPFS and ROMFS), sounds good.